### ISSUE TITLE
`yarn pack` produces slightly different `.tgz` files on different operating systems

### ISSUE THREAD

**Issue by borekb** [yarnpkg/berry#2774]
_Creating an issue from [this Discord chat](https://discord.com/channels/226791405589233664/226791405589233664/833651650631172098)._

**Describe the bug**

Running `yarn pack` produces slightly different archive (`.tgz` file) depending on the OS. For example, this is macOS vs. Windows: 

<img width="1121" alt="Screen Shot 2021-04-19 at 14 50 25" src="https://user-images.githubusercontent.com/101152/115239122-b1434a00-a11e-11eb-802c-ff88916cd455.png">

It comes down to the `OS` header, as described in [RFC1952](https://www.rfc-editor.org/rfc/rfc1952.html#page-7).

**To Reproduce**

1. Run `yarn pack` on e.g. macOS.
1. Run `yarn pack` on e.g. Windows.
1. Compare the files.

**Environment if relevant (please complete the following information):**

 - OS: combination of OS's, tested on macOS and Windows.
 - Node version: 12 & 14
 - Yarn version: 2.4.1

**Additional context**

[Discord chat](https://discord.com/channels/226791405589233664/226791405589233664/833651650631172098).

_Search terms: operating system, OS, Windows, Linux, macOS, Mac, `yarn pack`, tarball, archive, TGZ, GZIP, ZIP._

**Comment by arcanis** [comment]
It turns out Node [encodes the operating system](https://github.com/nodejs/node/issues/18273) when compressing things with the native zlib, so different systems will have different checksums. Fortunately it hasn't been a problem with the cache because we now use zip w/ a wasm version of the zlib (thus guaranteeing determinism), but `yarn pack` uses `node-tar`, which uses the native primitive for compression.

Unfortunately Node doesn't expose options to control this behaviour and, even if they did, it probably wouldn't be enough: parallelism may cause other deltas during compression, and there would be no way to control that either. The right fix would probably be to use the wasm zlib for compression (thankfully performances for `yarn pack` aren't a huge concern since it's only uncommonly used), but it's a bit of work 🤔

**Comment by borekb** [comment]
Does `npm pack` suffer from the same issue or do they avoid it somehow? (I don't have a good way to test it right now, sorry.)

**Comment by arcanis** [comment]
Me neither 😄 I suspect they do; I remember someone mentioning other inconsistent checksum issues on pack when adding npm-powered packages as dependencies, a fews months ago. Perhaps it was this.

**Comment by mpetrunic** [comment]
It would seem this is causing checksum mismatch bugs when installing git dependencies as they get packed instead of used as it so it's pretty severe as it's causing everyone to ignore or update checksum mismatches.

**Comment by AndreKR** [comment]
Yeah, that's the workaround. I have a git dependency (`git+ssh://git@gitlab.example.com/foo/bar.git`) in my package.json and if I run `yarn install` on Windows and commit `yarn.lock` I have to run the CI (Linux) with `YARN_CHECKSUM_BEHAVIOR=update` otherwise I get `The remote archive doesn't match the expected checksum`.

**Comment by KholdStare** [comment]
Any updates? We're being bit by issue https://github.com/yarnpkg/berry/issues/5136, which is a result of this issue.

**Comment by alexanderkjeldaas** [comment]
This is making it impossible to create deterministic results with `yarn pack`.  Is there a workaround?
